### ISSUE TITLE
How dynamically trigger move of selected point in color-picker:sv element - n points up, left, down or right?

### ISSUE THREAD

**Issue by bongobongo** [taufik-nurrohman/color-picker#71]
I would like to make it possible, in custom javascript, to move the selected round point in color-picker:sv element - using e.g. the up, down, left and right arrows (and/or using the I, M, J and L keys) on the keyboard. 
And I would like to make it so the selected point moves n points in a given direction on each press of the arrow - 
where the value of n is in relation to how many points there are in total horisontally (and vertically). 

How do i get number of points in the color-picker:sv element - horisontally and vertically?
How do I move the selected point n points up, right, left or down - and trigger the same events that are triggered when I left click to select a color?

**Comment by taufik-nurrohman** [comment]
So, I just took out the `P2RGB()` and `RGB2HSV()` functions from the core, and this seems to work.

``` js
const picker = new CP(document.querySelector('input'));

function P2RGB(a) {
    let h = +a[0],
        s = +a[1],
        v = +a[2],
        r,
        g,
        b,
        i,
        f,
        p,
        q,
        t;
    i = Math.floor(h * 6);
    f = h * 6 - i;
    p = v * (1 - s);
    q = v * (1 - f * s);
    t = v * (1 - (1 - f) * s);
    i = i || 0;
    q = q || 0;
    t = t || 0;
    switch (i % 6) {
        case 0:
            r = v, g = t, b = p;
            break;
        case 1:
            r = q, g = v, b = p;
            break;
        case 2:
            r = p, g = v, b = t;
            break;
        case 3:
            r = p, g = q, b = v;
            break;
        case 4:
            r = t, g = p, b = v;
            break;
        case 5:
            r = v, g = p, b = q;
            break;
    }
    return [Math.round(r * 255), Math.round(g * 255), Math.round(b * 255)];
}

function RGB2HSV(a) {
    let r = +a[0] / 255,
        g = +a[1] / 255,
        b = +a[2] / 255,
        max = Math.max(r, g, b),
        min = Math.min(r, g, b),
        h,
        s,
        v = max,
        d = max - min;
    s = max === 0 ? 0 : d / max;
    if (max === min) {
        h = 0; // Achromatic
    } else {
        switch (max) {
            case r:
                h = (g - b) / d + (g < b ? 6 : 0);
                break;
            case g:
                h = (b - r) / d + 2;
                break;
            case b:
                h = (r - g) / d + 4;
                break;
        }
        h /= 6;
    }
    return [h, s, v];
}

picker.on('change', function(r, g, b, a) {
    this.source.value = this.color(r, g, b, a);
});

const step = .05; // Adjust the color step here!

picker.source.addEventListener('keydown', function(e) {
    let key = e.key,
        color = picker.get();
    let a = color.pop(); // Detach alpha value from the color list
    let [h, s, v] = RGB2HSV(color);
    if ('ArrowDown' === key) {
        v -= .05;
        v = v < 0 ? .01 : v; // Cannot use `0` because it will be normalized to red :(
        picker.set(...[...P2RGB([h, s, v]), a]);
        e.preventDefault();
    } else if ('ArrowLeft' === key) {
        s -= .05;
        s = s < 0 ? .01 : s; // Cannot use `0` because it will be normalized to red :(
        picker.set(...[...P2RGB([h, s, v]), a]);
        e.preventDefault();
    } else if ('ArrowRight' === key) {
        s += .05;
        s = s > 1 ? 1 : s;
        picker.set(...[...P2RGB([h, s, v]), a]);
        e.preventDefault();
    } else if ('ArrowUp' === key) {
        v += .05;
        v = v > 1 ? 1 : v;
        picker.set(...[...P2RGB([h, s, v]), a]);
        e.preventDefault();
    }
}, false);
```

No need to add special method to move the cursor by arrow keys. Changing the saturation and value is enough.

**Comment by bongobongo** [comment]
That was quick!!!
Excellent work.
Just changed this: 
v -= .05;
to
v -= step;
And similar change for the other three directions, and it worked with the step value.

Now I can probably write the code for moving it diagonally myself.

Much appreciated!!!!

**Comment by taufik-nurrohman** [comment]
Tha `<i>` tag movement was faked and does not have any event listener attached to it. Instead, the event listeners were attached to the container. The `<i>` tag then moved along based on the pointer offset relative to the container. It was only for presentational purpose.

**Comment by bongobongo** [comment]
Done some more testing and found a couple of more issues. 
Added them to the issue comment above - so it's all in same spot.

**Comment by bongobongo** [comment]
Just tested it. 
Found one issue: 
1. click inside hue. 
2.  Move indicator up with arrow key - to see color change in saturation area. 
3. then, when click inside saturation area, hue is reset to where it was on last click in hue area.
 It should stay put. 
 
 and this was missing: 
 if (tabIndex == 2) // this is when alpha pane is removed
     tabIndex = 0;
     
Except for this - it really works very well

**Comment by taufik-nurrohman** [comment]
[This](https://github.com/taufik-nurrohman/color-picker/commit/087c8daabf3a52785840da4b9785effeda07024f) should fix the hue/opacity bug.

**Comment by bongobongo** [comment]
Tested it again. 
Last fix introduced two issues: 
After changing hue with up or down key -
1) when using arrow keys in saturation, the movement is no longer on a straight line. 
2) when using arrow keys in saturation, the indikator in hue is sometimes moved up or down (not much) but should stay put.

The issues can be seen here: 
https://taufik-nurrohman.js.org/color-picker/tweak/keyboard.html

**Comment by taufik-nurrohman** [comment]
That&rsquo;s because of the color rounding on every key-up, caused the pointers to be re-positioned. What browser are you using? I am using Firefox and the difference is not that much.

Actually, I can fix this using debounce function but I am afraid that this solution won&rsquo;t last long.

Can you confirm that the movement will be stright when you keep pressing the arrow key?

**Comment by taufik-nurrohman** [comment]
> When it reaches bottom of saturation - the marker is moved to bottom left.

Yes, because there is no point in moving the pointer to the left or right at the bottom of the saturation. It will always be black.

**Comment by taufik-nurrohman** [comment]
I&rsquo;m sure the source that causes the eratic movement comes from `RGB2HSV` function. But `picker.set()` function only look at the provided RGB input. And this kind of input does not have any information about the direction in the UI.

Including the hue direction: both 0 and 100% have the same red color. Because hue is a circle and so 0° is equal to 360°. If we, for example put the hue slider to the top to get a red color and assuming that normally the default red placement is at the bottom, then if we refresh the color picker, it is likely the pointer will be placed at bottom because the color converter only understand about the input color. And input color does not have directions.

**Comment by taufik-nurrohman** [comment]
I am afraid that I cannot fix the buggy pointer position at the bottom unless I use some hack to store the exact pointer position somewhere. But I can try to find a better &ldquo;RGB to HSV&rdquo; converter function.

By the way, if you ever know other JavaScript color picker application that has keyboard navigation support, please let me know. Maybe I can implement some feature from there for this tweak.

**Comment by taufik-nurrohman** [comment]
> before this last "fix" - using arrow down, did not make marker change direction nor to pop to a bottom corner.

If you remove the `source.addEventListener('keyup', function(e) {})` part, the movement bug will not happen, but the internal color value will not update.

**Comment by bongobongo** [comment]
I wrote this some comments back, after you introduced an enhansed version: 
"Just tested it.
Found one issue:

click inside hue.
Move indicator up with arrow key - to see color change in saturation area.
then, when click inside saturation area, hue is reset to where it was on last click in hue area.
It should stay put..."

I looked a bit more on this version, and for me - the most imporant thing is that it is "pixel perfect" in the saturation area - when you use the arrow keys. 
That version accomplish that goal. 
**The only issue with that version was that navigating inside the hue area using arrows, after clicking on it, caused a colorshift in saturation when saturation was cllicked.**

But after testing it a few days now, I realize that it's not very important to be able to use arrow keys to select the hue. If I want to select a specific hue, then I can allways enter a hex code in the input field and hit enter (added an onenter event to input field). 

**So - perhaps - remove arrow navigation for hue and alpha for that version - and then we have something that is fully functional - and "pixel perfect".**

**Comment by taufik-nurrohman** [comment]
Here with exact click event.

https://user-images.githubusercontent.com/1669261/136717970-f36ced81-f798-4a11-9125-52b7313d3e38.mp4

**Comment by taufik-nurrohman** [comment]
Sigh! Found the issue after dragging.

https://user-images.githubusercontent.com/1669261/136718320-0bde92aa-e8be-48d2-bc28-b409378c97e6.mp4
